RecoEcal/EgammaClusterAlgos/src/EndcapPiZeroDiscriminatorAlgo.cc

```diff
@@ -104,15 +104,15 @@ vector<float> EndcapPiZeroDiscriminatorAlgo::findPreshVector(ESDetId strip,  Rec
   final_strip--;
   ESDetId last_stripID = final_strip->first;
 
-  float E = 0;
   vector<ESDetId>::iterator itID;
   for (itID = road_2d.begin(); itID != road_2d.end(); itID++) {
     LogTrace("EcalClusters") << "EndcapPiZeroDiscriminatorAlgo: findPreshVectors: ID = " << *itID ;
   
+    float E = 0.;
     RecHitsMap::iterator strip_it = rechits_map->find(*itID);
```

Comment:
Looking at this my suggestion would also work:
```
-    if(goodPi0Strip(strip_it,last_stripID)) { // continue if strip not found in rechit_map
+    if ((strip_it != rechits_map->end()) && goodPi0Strip(strip_it,last_stripID)) { // continue if strip not found in rechit_map
```

If `rechits_map->find(*itID);` returns `rechits_map->end()` then call to `goodPi0Strip` would return `false`. Then you can actually delete some not needed lines. E.g., then you don't need `last_stripID`. That would overcomplexity here.

Reply:
Yes, that also works (and I have nothing against implementing it instead)

The difference is that if rechits_map->find(*itID); returns rechits_map->end(), then with it goodPi0Strip(...) wouldn't even be evaluated, thus not providing the LogTraces which are now in the code.

About the possibility of deleting not needed lines, this goes with my comment in the description of this PR: "I didn't want to modify the logic of the class/method, although I'm sure quite a lot can be done to improve it". 

I think this fix addresses the issue reported by you in #21173 with the minimal modification to the code currently in CMSSW. To remove the comparison between the last element of vector<ESDetId> road_2d and the first entry of the last element of the RecHitsMap *rechits_map (which I am also convinced as you that it is completely useless, and could be replaced by the very same check that you propose) I think one should first ask the authors of this code why they opted for such a convoluted logic instead, and make sure that there was no hidden particular case to be dealt with with it.

Reply:
You are responsible for reconstruction thus it's your call :) As long as it resolves the bug I am happy. Cleaning up the code would be nice-to-have.